src/python/pants/engine/internals/build_files.py

```diff
@@ -184,6 +188,49 @@ async def parse_address_family(
         )
         for fc in digest_contents
     ]
+
+    # Expunge/extend synthetic target adaptors with targets declared in BUILD files with the same
+    # address.
+    for address_map in address_maps:
+        for name, target_adaptor in address_map.name_to_target_adaptor.items():
+            if name not in synthetic_address_map.name_to_target_adaptor:
+                continue
+
+            # Pop synthetic target, as it will live in the BUILD file family.
+            synthetic_target_adaptor = synthetic_address_map.name_to_target_adaptor.pop(name)
+
+            # Check target type matches.
+            if synthetic_target_adaptor.type_alias != target_adaptor.type_alias:
```

Comment:
I hope this isn't too aggressive. Probably not? 

The alternative would be to only error if `_extend_synthetic` is True, and just silently accept the user's target definition otherwise?

Reply:
Yeah, I think you have a good point, if I can't change the type of the synthesized target, it is obtrusive.